// mcl-wasm@1.4.0/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mod = exports.curveType = exports.toHexStr = exports.toHex = exports.free = exports.fromHexStr = exports.init = exports.getRandomValues = exports.EC_PROJ = exports.IRTF = exports.NIST_P256 = exports.NIST_P224 = exports.NIST_P192 = exports.SECP384R1 = exports.SECP256K1 = exports.SECP224K1 = exports.BLS12_381 = exports.BN_SNARK1 = exports.BN462 = exports.BN381_2 = exports.BN381_1 = exports.BN254 = void 0;
var constants_1 = require("./constants");
Object.defineProperty(exports, "BN254", { enumerable: true, get: function () { return constants_1.BN254; } });
Object.defineProperty(exports, "BN381_1", { enumerable: true, get: function () { return constants_1.BN381_1; } });
Object.defineProperty(exports, "BN381_2", { enumerable: true, get: function () { return constants_1.BN381_2; } });
Object.defineProperty(exports, "BN462", { enumerable: true, get: function () { return constants_1.BN462; } });
Object.defineProperty(exports, "BN_SNARK1", { enumerable: true, get: function () { return constants_1.BN_SNARK1; } });
Object.defineProperty(exports, "BLS12_381", { enumerable: true, get: function () { return constants_1.BLS12_381; } });
Object.defineProperty(exports, "SECP224K1", { enumerable: true, get: function () { return constants_1.SECP224K1; } });
Object.defineProperty(exports, "SECP256K1", { enumerable: true, get: function () { return constants_1.SECP256K1; } });
Object.defineProperty(exports, "SECP384R1", { enumerable: true, get: function () { return constants_1.SECP384R1; } });
Object.defineProperty(exports, "NIST_P192", { enumerable: true, get: function () { return constants_1.NIST_P192; } });
Object.defineProperty(exports, "NIST_P224", { enumerable: true, get: function () { return constants_1.NIST_P224; } });
Object.defineProperty(exports, "NIST_P256", { enumerable: true, get: function () { return constants_1.NIST_P256; } });
Object.defineProperty(exports, "IRTF", { enumerable: true, get: function () { return constants_1.IRTF; } });
Object.defineProperty(exports, "EC_PROJ", { enumerable: true, get: function () { return constants_1.EC_PROJ; } });
var getRandomValues_1 = require("./getRandomValues");
Object.defineProperty(exports, "getRandomValues", { enumerable: true, get: function () { return __importDefault(getRandomValues_1).default; } });
var mcl_1 = require("./mcl");
Object.defineProperty(exports, "init", { enumerable: true, get: function () { return mcl_1.initializeMcl; } });
Object.defineProperty(exports, "fromHexStr", { enumerable: true, get: function () { return mcl_1.fromHexStr; } });
Object.defineProperty(exports, "free", { enumerable: true, get: function () { return mcl_1.free; } });
Object.defineProperty(exports, "toHex", { enumerable: true, get: function () { return mcl_1.toHex; } });
Object.defineProperty(exports, "toHexStr", { enumerable: true, get: function () { return mcl_1.toHexStr; } });
Object.defineProperty(exports, "curveType", { enumerable: true, get: function () { return mcl_1.initializedCurveType; } });
__exportStar(require("./value-types"), exports);
/** @internal for only testing */
var mcl_2 = require("./mcl");
Object.defineProperty(exports, "mod", { enumerable: true, get: function () { return mcl_2.mod; } });
